src/coreclr/jit/morph.cpp

```diff
@@ -371,24 +587,42 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
                 switch (dstType)
                 {
                     case TYP_INT:
+#ifdef TARGET_XARCH
+                        if (!tree->IsSaturatedConversion())
+                        {
+                            return fgMorphCastIntoHelper(tree, CORINFO_HELP_DBL2INT, oper);
+                        }
+#endif // TARGET_XARCH
                         return nullptr;
 
                     case TYP_UINT:
-#if defined(TARGET_ARM) || defined(TARGET_AMD64)
+#if defined(TARGET_ARM)
                         return nullptr;
-#else  // TARGET_X86
+#elif defined(TARGET_XARCH)
+                        if (tree->IsSaturatedConversion())
+                        {
+                            return nullptr;
+                        }
```

Comment:
These still seem unnecessary to me. We shouldn't be double morphing nodes and already have debug checks to validate that.

I'd expect that our normal flow should already only process them once and then be done, with no need to separately track if any saturating logic has been added or not.

Reply:
But at certain places, when I dont recurse, the node is morphed again and it hits the assert. https://github.com/dotnet/runtime/pull/97529#discussion_r1533097268

Reply:
I expect there is an issue with how the nodes are being introduced or marked then, as we don't have this issue in other places in morph.

There are many places that we insert new trees or modify existing ones and the normal expectation is that we should simply mark them with `INDEBUG(tree->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);` and then return the tree rather than recursively calling `fgMorphTree`.

It's likely important to find where the recursive morph call is occurring as that will pinpoint the problem, it's likely part of this PR since we don't hit any recursion issues in release mode where no tracking of whether a morph happened exists.

Reply:
Hi @tannergooding ,
Looks like in this case, simply marking the node with ```GTF_DEBUG_NODE_MORPHED``` does not work. This is because, some cases in the newly added code end up implementing the code path for ```CORINFO_HELP_DBG_IS_JUST_MY_CODE``` which is a ```GT_CALL```. Upon marking the new nodes morphed and simply returning the tree throws the following error -->
```
  D:\Git_repos\runtime\.dotnet
  Missing flags on tree [000001]: 10
  6
  99
  --X-------


  D:\Git_repos\runtime\src\coreclr\jit\fgdiagnostic.cpp:3555
  Assertion failed '!"Missing flags on tree"' in 'System.Half:op_Explicit(System.Half):uint' during 'Post-Morph' (IL size 9; hash 0x039992f4; FullOpts)
```

I have added some debug prints here https://github.com/khushal1996/runtime/blob/417ea5306a7faade39c53c60f8f5cd7deb816741/src/coreclr/jit/fgdiagnostic.cpp#L3565C1-L3570C26 and looks like the GT_CALL does not have correct flag when the node is simply marked morphed. Any ideas which area of the code can resolve this? Also, I have simplified the recursion by removing the ```do {} while(false)``` block as well as simplifying the code for executing helper functions. I tried adding few debug prints, debugging along the code where ```CORINFO_HELP_DBG_IS_JUST_MY_CODE``` is executed and how these flags are decided. I did not find a reason as to why marking the nodes as morphed is not sufficient. Can you help? Or do you think leaving the code in current state is better?

Reply:
So basically it goes into "JustMyCode" regardless of it is marked morphed or not.

Reply:
I would expect we do need and want to do the transform if we are replacing ourselves with a call and there should be no concern of recursive morph in that scenario (we're removing the `GT_CAST` and replacing it entirely with a `GT_CALL`).

The place where we don't want to recursively call morph is when we're finding `GT_CAST` and inserting new nodes prior to it, keeping the `GT_CAST` as is. In this scenario, recursion would just incur indefinitely and continuously add the new nodes prior to it, which would be bad. -- This is also where long term I think inserting that logic in lowering is better, as it avoids the entire issue and keeps the HIR simple/intuitive.

Reply:
Makes sense. But in cases where we cannot keep the ```GT_CAST``` intact and have to remove it and insert new nodes + new cast for e.g. https://github.com/khushal1996/runtime/blob/bb8a6e7bccf4046b34e9228063ad9eacfd9807ca/src/coreclr/jit/morph.cpp#L505C1-L557C10
```
        else if (dstType == TYP_UINT && compOpportunisticallyDependsOn(InstructionSet_SSE41))
        {
            if (srcType == TYP_FLOAT)
            {
                oper      = gtNewCastNode(TYP_DOUBLE, oper, false, TYP_DOUBLE);
                srcType   = TYP_DOUBLE;
                fieldType = CORINFO_TYPE_DOUBLE;
            }
            ssize_t     actualMaxVal  = UINT32_MAX;
            CorInfoType destFieldType = CORINFO_TYPE_LONG;

            oper               = gtNewSimdCreateBroadcastNode(TYP_SIMD16, oper, fieldType, 16);
            GenTree* op1Clone1 = fgMakeMultiUse(&oper);
            GenTree* op1Clone2 = fgMakeMultiUse(&oper);
            GenTree* op1Clone3 = fgMakeMultiUse(&oper);
            GenTree* op1Clone4 = fgMakeMultiUse(&oper);

            // get the max/min value vector
            GenTree* minVal    = gtNewDconNodeD(static_cast<double>(0));
            minVal             = gtNewSimdCreateBroadcastNode(TYP_SIMD16, minVal, fieldType, 16);
            GenTree* maxVal    = gtNewDconNodeD(static_cast<double>(actualMaxVal));
            maxVal             = gtNewSimdCreateBroadcastNode(TYP_SIMD16, maxVal, fieldType, 16);
            GenTree* maxValDup = fgMakeMultiUse(&maxVal);

            // check NaN
            GenTree* mask1 = gtNewSimdCmpOpNode(GT_EQ, TYP_SIMD16, oper, op1Clone1, fieldType, 16);
            // check negative
            GenTree* mask2 = gtNewSimdCmpOpNode(GT_GE, TYP_SIMD16, op1Clone2, minVal, fieldType, 16);

            // and mask
            GenTree* mask12 = gtNewSimdBinOpNode(GT_AND, TYP_SIMD16, mask1, mask2, fieldType, 16);

            // inp = inp & mask
            GenTree* saturated_val   = gtNewSimdBinOpNode(GT_AND, TYP_SIMD16, op1Clone3, mask12, fieldType, 16);
            GenTree* saturatedValDup = fgMakeMultiUse(&saturated_val);

            // compare with max value of uint
            GenTree* mask3 = gtNewSimdCmpOpNode(GT_GE, TYP_SIMD16, saturated_val, maxVal, fieldType, 16);

            // Select based on mask3
            saturated_val = gtNewSimdCndSelNode(TYP_SIMD16, mask3, maxValDup, saturatedValDup, fieldType, 16);
            // saturated_val = fgMorphTree(saturated_val);

            // scalar
            GenTree* val = gtNewSimdHWIntrinsicNode(srcType, saturated_val, NI_Vector128_ToScalar, fieldType, 16);
            val          = fgMorphTree(val);

            // cast it
            tree = gtNewCastNode(TYP_INT, val, false, dstType);
            INDEBUG(tree->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);
            // tree->gtFlags &= (~GTF_EXCEPT & ~GTF_CALL);
            return tree;
```
Here, if we dont call ```val = fgMorphTree(val)```[link](https://github.com/khushal1996/runtime/blob/bb8a6e7bccf4046b34e9228063ad9eacfd9807ca/src/coreclr/jit/morph.cpp#L550), we hit the assert in morph ```did not morph this node```. According to you, is this a coding structure problem or something in the logic where I am adding new nodes? According to you, in the above code, does marking ```INDEBUG(tree->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);``` should be enough and there should be no need to morph the operands?

Reply:
I think I dont understand your comment. Mainly, if we find a ```GT_CAST``` node, and we change the nodes prior to it, we dont want to remorph it since it will lead to recursion; this is what I undestood from your comment. But what about the nodes we added after the ```GT_CAST```? Won't we need to morph those? Just marking the ```GT_CAST``` as morphed is not sufficient it seems. Neither is marking the end result as morphed sufficient. I hope I am making sense in my comment. Can you give a code flow example of what you are suggesting?

What the code look right now is -->
```
GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
{
    GenTree*  oper    = tree->CastOp();
    if (//certain conditions)
    {
        //modify oper
        //tree->AsCast()->CastOp() = newOper;
        //cnd1 = make conditional select on tree and maxVal
        // return fgMorphTree(cnd1);
    }
}
```

Reply:
I might need to debug into it to understand the issue you're hitting.

In general, morph is expected to be called exactly once per node and that's why `GTF_DEBUG_NODE_MORPHED` is debug only, because it's just meant to be a safety check. Having to introduce a new flag that indicates whether a cast has been morphed and the relevant "saturating" logic was introduced on top is somewhat counter-intuitive to that process and is why I think it should be unnecessary/undesirable.

For cases like `CALL` there are complex changes that morph might do, including correctly fixing up arguments to match the ABI and so if we replace a `GT_CAST` with a `GT_CALL` we do need to call `fgMorphTree` on it, to ensure that it has all the relevant transforms done (this morph call will itself then mark the call node with `GTF_DEBUG_NODE_MORPHED`).

For other nodes we introduce, we similarly need to either call `fgMorphTree` or manually mark them with `INDEBUG(tree->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);`. The only case that cannot call `fgMorphTree` again is any tree that contains a `GT_CAST` which we've already morphed.

The latter issue there, not being able to call `fgMorphTree` on a tree that contains a `GT_CAST` we've already morphed is what you're currently handling via the `IsSaturatedConversion()` flag, but which I'm suggesting we should be able to handle differently. 

In particular, we know that there aren't any transforms that would happen for things like `gtNewSimdCmpOpNode` or `gtNewSimdCndSelNode`, so we should be able to directly mark them via `INDEBUG(tree->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);`. The issue of, what if `op1` is or contains a `GT_CAST` node itself is generally not a concern because of how morph works (`fgMorphTree(node)` is responsible for calling `fgMorphTree` on any children of `node`, so we're the only one's morphing `op1`, leading to no chance for infinite recursion or double morphing unless our code is doing something wrong)

Long term however, and maybe it would even be simpler at this point to just fix it in the PR; this logic should probably be in lowering instead. This is because it avoids this issue from being hit entirely (morph only has to be concerned with replacing a `CAST` with a `CALL` to the JIT helper if we don't have lowering handling for it). That also avoids any TP concerns from larger HIR and from some other phase introducing a cast that isn't correctly fixed up to include the saturation (as it's always simply inserted on the backend when we're deciding what instructions to emit).